### PR TITLE
Trust based on EKPub

### LINKED ISSUE #4: Trust based on EKPub
**Describe the solution you'd like**

Currently, EKCert chain must be validated and a node is allowed.  This requires the server to have CA roots for each TPM manufacturer, and trusts all devices with a matching CA root.

I think it would also be beneficial to allow for a list of EKPub hashes to be considered.  This would mean trust could be locked down to only certain devices that are on the EKPub whitelist

One solution would be an optional server configuration, such as:

```
ek_pub_hash_path  = "/opt/spire/.data/ek_pub_hashes"
```

Each file in that directory would be an empty file with a filename being a SHA-256 hash of an authorized EK Pub.  If `ek_pub_hash_path` is supplied, a check for file exists on `{ek_pub_hash_path}/{ek_pub_hash}` would be performed and if found, considered valid.

**Describe alternatives you've considered**

Considered a datastore API for checking EK Pubs that reside in a remote datastore, but that would introduce additional complexity into this project, and could be handled out-of-band.

**Additional context**

The solution would be similar to the "Trust based on EKPub" described here: https://docs.microsoft.com/en-us/windows-server/identity/ad-ds/manage/component-updates/tpm-key-attestation#BKMK_DeploymentOverview

### LINKED ISSUE #7: Use hex-encoded Public Key Hash instead of Base64 to avoid / char in SPIFFE-ID
[SPIFFE-ID paths](https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE-ID.md#22-path) may be hierarchial, and using the base64 encoded Public Key Hash results in SVIDs that look like:

`spiffe://domain.test/spire/agent/tpm/c1XbxrikL+sgdCw7hKFlmh2/y/QenHiIfLazBlsG/yQ=`

This appears as if the path is hierarchial, since base64 includes `/` in it's character set

Additionally, #4 proposes storing authorized Public Key Hashses as an empty file on the filesystem.  Since `/` is a directory separator on Linux, it cannot be represented in a file

Hex encoding for SHA256 representations is fairly standard practice, and would alleviate these issues

### PR DESCRIPTION
**Pull request by caleblloyd**
- Fixes #4 
- Fixes #7 

**Describe your changes**

1. Add trust based on EK Hash, example:

```bash
agent  $ ./get_tpm_pubhash
1b5bbe2e96054f7bc34ebe7ba9a4a9eac5611c6879285ceff6094fa556af485c 

server $ mkdir -p /opt/spire/.data/hashes
server $ touch /opt/spire/.data/hashes/1b5bbe2e96054f7bc34ebe7ba9a4a9eac5611c6879285ceff6094fa556af485c
```

2. Handle TPMs that only contain EK Public Keys and not EK Certificates.  Agent PEM Encodes either X509 Certificate or PKIX Public Key (which supports RSA and EC keys, which are both valid for TPM).  Server reads PEM block header to determine which type was sent.

3. Change EK Pub Hash to Hex Encoded sha256 hash of PKIX Public Key bytes

4. Update spire dependencies and test files to `0.10`

5. Add a `./ci/pack.sh` script to pack a release tarball conaining `tpm_attestor_agent`, `tpm_attestor_server`, and `get_tpm_pubhash` binaries

6. Add documentation to `README.md`

**Testing performed**

1. EK Hash tested comprehensively alongside CA Cert in `pkg/agent/plugin_test.go`

2. Certificate / Public Key encoding and decoding tested in `pkg/common/tpm_attestor_test.go`

3. EK PubHash encoding tested in `tools/get_tpm_pubhash/get_tpm_pubhash_test.go`

### CODE REVIEW

File: pkg/common/tpm_attestor_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	testCases := []struct {
+		name     string
+		hasCert  bool
+		pemBytes []byte
+		pubHash  string
+	}{
+		{
+			name:    "certificate",
+			hasCert: true,
+			pemBytes: []byte(`-----BEGIN CERTIFICATE-----
```
Comment (pawalt) [inline pkg/common/tpm_attestor_test.go:29]: Could you add some notes on how to generate this test data? Can be top-level docs in the README or comments above the places where test data exists.

**Review by pawalt — CHANGES_REQUESTED** [review]

File: pkg/common/tpm_attestor_test.go (hunk 1 above)
Comment (caleblloyd) [reply pkg/common/tpm_attestor_test.go:29]: :+1: added openssl commands as comments

**Comment by twexler** [comment]
Hey @caleblloyd, this looks okay from a code perspective, I'm trying to get ramped up on how to test this on an actual machine to verify that everything seems to be copacetic. I'll keep you in the loop as I make progress.

**Comment by caleblloyd** [comment]
To test on hardware, you can comment out `CommandChannel` in the `OpenTPM` calls within the `*_test.go` files and call `go test`

Then change the expected PubHash and manufacturer CA cert to match your hardware. Example:

```
tpm, err := attest.OpenTPM(&attest.OpenConfig{
		TPMVersion:     attest.TPMVersion20,
		// CommandChannel: &common_test.TPMCmdChannel{ReadWriteCloser: s},
	})
```

**Comment by rgl** [comment]
Any progress on this?

**Comment by caleblloyd** [comment]
@rgl our fork is up-to-date with this change - https://github.com/boxboat/spire-tpm-plugin

**Comment by kfox1111** [comment]
ping

**Comment by dennisgove** [comment]
@pawalt are you comfortable with these changes?
@twexler do you have some time to a look at this?

**Comment by dennisgove** [comment]
This looks very good. I just want to confirm that the addition here is support of EK Public Keys via their hashes. This adds an alternative for TPMs where EK Certificates are not available?

**Review by dennisgove — APPROVED** [review]
lgtm.

**Comment by dennisgove** [comment]
Thank you for this enhancement, @caleblloyd. I'm very sorry for the delay in handling it.

**Comment by caleblloyd** [comment]
Oh neat, thanks!  I think some more folks have updated the plugin on this fork also: https://github.com/boxboat/spire-tpm-plugin

**Pull request merged**